src/server/pegasus_service_app.h

```diff
@@ -57,11 +59,13 @@ class pegasus_replication_service_app : public ::dsn::replication::replication_s
         ::dsn::error_code ret = ::dsn::replication::replication_service_app::stop();
         if (_updater_started) {
             pegasus_counter_reporter::instance().stop();
+            _builtin_metrics.stop();
         }
         return ret;
     }
 
 private:
+    dsn::builtin_metrics _builtin_metrics;
     bool _updater_started;
```

Comment:
If pegasus_counter_reporter and builtin_metrics are both common objects, and they can stop automatically, _updater_started can be removed?

Reply:
`service_app::start(...)` (such as `dsn::replication::replication_service_app::start(args_new)`) will return an error code. Once this error code is not `ERR_OK`, metric service will not be started. That's why `_updater_started ` is involved.

We can do another CHECK for the returned error code by `service_app::start(...)`. Actually the root caller, `start_app()` in service_control_task::exec() will also do CHECK for `ERR_OK`. Do CHECK here to guarantee that all following services (such as built-in metrics) are started, then `_updater_started ` could be removed.